pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/BaseRoutingTableBuilder.java

```diff
@@ -86,10 +148,52 @@ protected void setRoutingTables(List<Map<String, List<String>>> routingTables) {
     return _routingTables;
   }
 
-  protected void handleNoServingHost(String segmentName) {
-    LOGGER.error("Found no server hosting segment {} for table {}", segmentName, _tableName);
-    if (_brokerMetrics != null) {
-      _brokerMetrics.addMeteredTableValue(_tableName, BrokerMeter.NO_SERVING_HOST_FOR_SEGMENT, 1);
+  /**
+   * Computes a routing table on-the-fly using a mapping of segment to servers. Because of the performance concern,
+   * the default behavior is to randomly pick a server among available servers for each segment.
+   *
+   * @param segmentsToQuery a list of segments that need to be processed for a particular query
+   * @return a routing table
+   */
+  public Map<String, List<String>> computeDynamicRoutingTable(Set<String> segmentsToQuery) {
+    Map<String, List<String>> segmentToServersMap = getSegmentToServersMap();
+    Map<String, List<String>> routingTable = new HashMap<>();
+
+    for (String segmentName : segmentsToQuery) {
+      List<String> servers = segmentToServersMap.get(segmentName);
+      if (servers.size() != 0) {
+        String selectedServer = servers.get(_random.nextInt(servers.size()));
+        List<String> segments = routingTable.get(selectedServer);
+        if (segments == null) {
+          segments = new ArrayList<>();
+          routingTable.put(selectedServer, segments);
+        }
+        segments.add(segmentName);
+      } else {
+        handleNoServingHost();
+      }
     }
+    return routingTable;
   }
+
+  /**
+   * Given an external view and a list of instance configs, computes the mapping of segment to servers. The mapping
+   * will be cached if we use dynamic routing.
+   *
+   * @param externalView an external view
+   * @param instanceConfigs a list of instance config
+   * @return a mapping of segment to servers
+   */
+  abstract Map<String, List<String>> computeSegmentToServersMapOnExternalViewChange(ExternalView externalView,
+      List<InstanceConfig> instanceConfigs);
+
+  /**
+   * Given a mapping of segment to servers, compute a list of final routing tables that will be cached when
+   * we use pre-computing routing.
+   *
+   * @param segmentToServersMap a mapping of segment to servers
+   * @return a list of final routing tables
+   */
+  abstract List<Map<String, List<String>>> computeRoutingTablesOnExternalViewChange(
```

Comment:
Rename this method to something like computeRoutingTablesFromSegmentToServersMap()

Reply:
done